src/java.desktop/share/classes/javax/swing/JEditorPane.java

```diff
@@ -473,68 +473,76 @@ public void setPage(URL page) throws IOException {
             throw new IOException("invalid url");
         }
         URL loaded = getPage();
-
+        Object postData = getPostData();
 
         // reset scrollbar
         if (!page.equals(loaded) && page.getRef() == null) {
-            scrollRectToVisible(new Rectangle(0,0,1,1));
+            scrollRectToVisible(new Rectangle(0, 0, 1, 1));
         }
-        boolean reloaded = false;
-        Object postData = getPostData();
-        if ((loaded == null) || !loaded.sameFile(page) || (postData != null)) {
-            // different url or POST method, load the new content
-
-            int p = getAsynchronousLoadPriority(getDocument());
-            if (p < 0) {
-                // open stream synchronously
-                InputStream in = getStream(page);
-                if (kit != null) {
-                    Document doc = initializeModel(kit, page);
-
-                    // At this point, one could either load up the model with no
-                    // view notifications slowing it down (i.e. best synchronous
-                    // behavior) or set the model and start to feed it on a separate
-                    // thread (best asynchronous behavior).
-                    p = getAsynchronousLoadPriority(doc);
-                    if (p >= 0) {
-                        // load asynchronously
-                        setDocument(doc);
-                        synchronized(this) {
-                            pageLoader = new PageLoader(doc, in, loaded, page);
-                            pageLoader.execute();
-                        }
-                        return;
-                    }
-                    read(in, doc);
-                    setDocument(doc);
-                    reloaded = true;
-                }
-            } else {
-                // we may need to cancel background loading
-                if (pageLoader != null) {
-                    pageLoader.cancel(true);
-                }
 
+        if ((postData == null) && (page.sameFile(loaded))) {
+            if (page.getRef() != null) {
+                scrollToReference(page.getRef());
+            }
+            return;
+        }
```

Comment:
```suggestion
        if ((postData == null) && (page.sameFile(loaded))) {
            // The same page with different reference
            if (reference != null) {
                scrollToReference(reference);
            } else {
                // Scroll to the top of the page
                scrollRectToVisible(new Rectangle(0, 0, 1, 1));
            }
            return;
        }
```
Here, `reference` contains the value of `page.getReference()`. I suggest moving that variable to the top too (after `postData`) and re-use it.

I also suggest marking `loaded`, `postData`, `reference` as `final`, it explicitly indicates the values are not changed in the method.